### DISCUSSION TITLE
Skip instruction implementation

### DISCUSSION THREAD

**Discussion by astrelsky** [NationalSecurityAgency/ghidra#4241] [Q&A] [answered]
I've been struggling to implement skip instructions for the RL78 for quite some time now. It seems the best way to go about it is to use a similar approach to what is used for ARM. However, because any instruction can be skipped, this makes all branches, calls and returns conditional and causes the decompiler to fail 90% of the time. This is usually due to exceeding the maximum instruction count.

An alternative is to use local context and this actually works great, until it doesn't. The problem here occurs when an instruction which follows a skip instruction is branched into. Since there is a difference in instruction context disassembly "fails" and the decompiler will usually start puking `BADSPACEBASE` everywhere.

I think the best way, other than built in support for skip instructions, would be to use the first approach with some way to tell the flow analysis and the decompiler that a branch/call/return is only conditional if it has been proceeded by a skip instruction. Any ideas on how to do this that won't suffer from the same issues when using context?

**Comment by astrelsky** [comment]
I think I just figured out a solution. I will use pcode injection for the skip instructions because I can get the next instruction and its lenses to inject the appropriate pcode to skip it from the java api.

**Reply by astrelsky** [reply]
So I setup the pcode inject library and the injection. This is what I have.

```java
	private String doGetCondition(char c) {
		switch (c) {
			case 'C':
				return "PSW[0,1]";
			case 'N':
				return "!";
			case 'Z':
				return "PSW[6,1]";
			case 'H':
				return "(PSW[6,1] | PSW[0,1])";
		}
		throw new AssertException();
	}

	private String getCondition(Program program, InjectContext con) {
		Instruction inst = program.getListing().getInstructionAt(con.baseAddr);
		String mnemonic = inst.getMnemonicString();
		try {
			String cond = doGetCondition(mnemonic.charAt(2));
			if (cond.equals("!")) {
				cond += doGetCondition(mnemonic.charAt(3));
			}
			return "if (" + cond + ") ";
		} catch (AssertException | IndexOutOfBoundsException e) {
			Msg.showError(this, null, mnemonic, e);
			throw e;
		}
	}

	private String getJumpTarget(Program program, InjectContext con) {
		Instruction inst = program.getListing().getInstructionAt(con.nextAddr);
		if (inst == null) {
			return "";
		}
		return "goto [inst_next+"+Integer.toString(inst.getLength())+":3];";
	}

	@Override
	public PcodeOp[] getPcode(Program program, InjectContext con) {
		SleighLanguage l = con.language;
		PcodeParser parser = new PcodeParser(l, l.getUniqueBase());
		String target = getJumpTarget(program, con);
		String cond = getCondition(program, con);
		String sourceName = getSource();
		ConstructTpl constructTpl = parser.compilePcode(cond + target, sourceName, 1);
		SleighParserContext protoContext =
			new SleighParserContext(con.baseAddr, con.nextAddr, con.refAddr, con.callAddr);
		ParserWalker walker = new ParserWalker(protoContext);
		PcodeEmitObjects emit = new PcodeEmitObjects(walker);
		try {
			emit.build(constructTpl, -1);
		} catch (Exception e) {
			throw new AssertException(e);
		}
		return emit.getPcodeOp();
	}
```

But this ends up resulting in the following
```
2022-05-12 20:09:23 ERROR (PcodeCompile) C:\Users\astre\Documents\ghidra_10.2_DEV\Ghidra\Extensions\RL78_sleigh\data\languages\rl78.cspec:1: invalid dynamic target used in goto destination  
2022-05-12 20:09:23 ERROR (DecompileCallback) Decompiling ram:01501a, pcode inject error at ram:015046: Unexpected Error: Semantics for this instruction are not implemented ghidra.util.exception.AssertException: Unexpected Error: Semantics for this instruction are not implemented

```

Anyone have any ideas? I don't see why this can't be done.

**Reply by astrelsky** [reply]
@caheckman or @ghidra1 any thoughts on why the following would result in `Low-level Error: Bad relative branch at instruction : (ram,0x01938c)`? This seems like it should be pretty trivial.

```java
public class InjectSkipPayload extends InjectPayloadCallother {

	public InjectSkipPayload(String sourceName, SleighLanguage language) {
		super(sourceName);
	}

	@Override
	public PcodeOp[] getPcode(Program program, InjectContext con) {
		Instruction inst = program.getListing().getInstructionAt(con.nextAddr).getNext();
		Varnode[] in = {new Varnode(inst.getAddress(), 3)};
		PcodeOp[] p = {new PcodeOp(con.baseAddr, -1, PcodeOp.BRANCH, in)};
		return p;
	}

}
```

The decompilation dump has the following for the injection.
```xml
<injectdebug>
<inject name="SKIP" type="2">
  <addr space="ram" offset="0x1938c"/>
  <payload><![CDATA[
<inst offset="2"><op code="4"><seqnum space="ram" offset="0x1938c"/><void/><addr space="ram" offset="0x19391" size="3"/></op></inst>

]]></payload>
</inject>
</injectdebug>
```

The instruction containing `SKIP` is implemented as follows
```
:SKZ is opcode=0x61; opcode=0xE8 {
    if $(Z) goto <skip>;
    goto <end>;
    <skip>
    SKIP();
    <end>
}
```

**Comment by astrelsky** [comment]
Well, this is extremely gross but here is a work around. Having to put up with the "removing unreachable block" warnings and error bookmarks everywhere is much much better than not having it implemented at all. Luckily the bookmarks from disassembly "error" isn't a problem because the disassembly by flow to the next instruction occurs first. The pcodeop `getNextInstructionLength` is implemented to do exactly what it's name implies via pcode injection.

```
define pcodeop getNextInstructionLength;

SKIP1: loc is epsilon [ loc = inst_next + 1; ] {
    export *[ram]:3 loc;
}

SKIP2: loc is epsilon [ loc = inst_next + 2; ] {
    export *[ram]:3 loc;
}

SKIP3: loc is epsilon [ loc = inst_next + 3; ] {
    export *[ram]:3 loc;
}

#
# Table 30/30
#

# SKC
:SKC is SKIP1 & SKIP2 & SKIP3 & opcode=0x61; opcode=0xC8 {
    len:2 = getNextInstructionLength();
    if ($(CY)) goto <skip>;
    goto inst_next;
    <skip>
    if (len == 1) goto SKIP1;
    if (len == 2) goto SKIP2;
    if (len == 3) goto SKIP3;
}
```

**Reply by astrelsky** [reply]
@emteere any thoughts on a more appropriate alternative? This does work it's just a bit hacky and creates a lot of noise with erroneous error bookmarks and "removing unreachable block" warnings.

**Reply by dev747368** [reply]
I'm a noob with sleigh stuff, and this is completely untested, but would modifying your goto logic to be more like whats in https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Processors/x86/data/languages/macros.sinc help?

**Reply by astrelsky** [reply]
> 

Unfortunately I don't think so. There doesn't seem to be any restriction as to what instruction can come after a skip instruction. I have seen `RET` come after one for example. These instructions can also be branched to so I can't used a context variable or a delay slot.

**Reply by GhidorahRex** [reply]
I think this is the best solution for now. Without knowing the next instruction length you can't implement a goto, and using context for the branch can clearly make the disassembly of the next instruction different. I'll try to mull it over some more.

**Reply by astrelsky** [reply]
> I think this is the best solution for now. Without knowing the next instruction length you can't implement a goto, and using context for the branch can clearly make the disassembly of the next instruction different. I'll try to mull it over some more.

I just had another hacky idea. I'm thinking about making an analyzer or a loader which would run before disassembly. It would go through and look for all possible skip instructions (there are only 5 I think). It would then use the pseudo disassembler to disassemble the next instruction (skip instructions are always 1 byte) and set a noflow context register value to the size of the next instruction. It's a bit more work but this should cover all everything. 


This processor is such a nightmare :sweat_smile:

**Reply by GhidorahRex** [reply]
it's always just one or two instructions that screws everything up

**Reply by astrelsky** [reply]
> it's always just one or two instructions that screws everything up

Yup. At least the problem here was obvious. I discovered yesterday that I had an instruction which was indirectly copying the stack pointer. Since it wasn't being handled explicitly it wasn't actually being detected as the stack pointer and was screwing everything up.

My latest hacky idea with the analyzer/loader seems to work best. Its just a bit weird because it'll cause a disassembly failure if the context register value was cleared or not set for some reason.

**Reply by GhidorahRex** [reply]
What about something like this:
```
define register offset=0x1010 size=1 [ SKIP_REG ];
define register offset=0x1000 size=4 contextreg;
define context contextreg
    skip_ctx=(0,2) noflow
;

:^instruction is instruction & skip_ctx=1 {
    local tmp:1 = SKIP_REG;
    SKIP_REG = 0;
    if (<skip_condition1> & tmp == 1) goto inst_next;
    build instruction;
}

:^instruction is instruction & skip_ctx=2 {
    local tmp:1 = SKIP_REG;
    SKIP_REG = 0;
    if (<skip_condition2> & tmp == 1) goto inst_next;
    build instruction;
}

:^instruction is instruction & skip_ctx=0 {
    build instruction;
}

:skip1 is bits=0xff [skip_ctx=1; globalset(inst_next,skip_ctx);] {
    SKIP_REG = 1;
}

:skip2 is bits=0xfe [skip_ctx=2; globalset(inst_next,skip_ctx);] {
    SKIP_REG = 1;
}

```

It uses a fake `SKIP_REG` register that is only set by the skip instruction, as well as a context register that specifies when an instruction needs to check if it comes right after a skip. Only the branches/conditionals that occur after a skip should be conditional, but the context won't ever change for those instructions so it shouldn't differ.

**Reply by astrelsky** [reply]
> What about something like this:
> ```
> define register offset=0x1010 size=1 [ SKIP_REG ];
> define register offset=0x1000 size=4 contextreg;
> define context contextreg
>     skip_ctx=(0,2) noflow
> ;
> 
> :^instruction is instruction & skip_ctx=1 {
>     local tmp:1 = SKIP_REG;
>     SKIP_REG = 0;
>     if (<skip_condition1> & tmp == 1) goto inst_next;
>     build instruction;
> }
> 
> :^instruction is instruction & skip_ctx=2 {
>     local tmp:1 = SKIP_REG;
>     SKIP_REG = 0;
>     if (<skip_condition2> & tmp == 1) goto inst_next;
>     build instruction;
> }
> 
> :^instruction is instruction & skip_ctx=0 {
>     build instruction;
> }
> 
> :skip1 is bits=0xff [skip_ctx=1; globalset(inst_next,skip_ctx);] {
>     SKIP_REG = 1;
> }
> 
> :skip2 is bits=0xfe [skip_ctx=2; globalset(inst_next,skip_ctx);] {
>     SKIP_REG = 1;
> }
> 
> ```
> 
> It uses a fake `SKIP_REG` register that is only set by the skip instruction, as well as a context register that specifies when an instruction needs to check if it comes right after a skip. Only the branches/conditionals that occur after a skip should be conditional, but the context won't ever change for those instructions so it shouldn't differ.

The instruction that follow a skip instruction is often branched into from elsewhere. This is the route I originally went but the difference in context from a branch and a skip caused problems.

The analyzer route seems to work nicely for now. I just need it to run automatically before instructions are disassembled and it currently needs to be run manually by selecting everything and using one shot analysis. 




Edit: Actually I misread it. That might work. I will try giving it a shot this weekend.

**Reply by astrelsky** [reply]
Sorry for the incoming mess, but this is what I tried and what I got.

```
# SKC
:SKC is opcode=0x61; opcode=0xC8 [skip_check = 1; globalset(inst_next, skip_check);] {
    SKIP_REG = $(CY);
}

# SKNC
:SKNC is opcode=0x61; opcode=0xD8 [skip_check = 1; globalset(inst_next, skip_check);] {
    SKIP_REG = !$(CY);
}

# SKZ
:SKZ is opcode=0x61; opcode=0xE8 [skip_check = 1; globalset(inst_next, skip_check);] {
    SKIP_REG = $(Z);
}

# SKNZ
:SKNZ is opcode=0x61; opcode=0xF8 [skip_check = 1; globalset(inst_next, skip_check);] {
    SKIP_REG = !$(Z);
}

# SKH
:SKH is opcode=0x61; opcode=0xE3 [skip_check = 1; globalset(inst_next, skip_check);] {
    SKIP_REG = ($(Z) | $(CY));
}

# SKNH
:SKNH is opcode=0x61; opcode=0xF3 [skip_check = 1; globalset(inst_next, skip_check);] {
    SKIP_REG = (!($(Z) | $(CY)));
}

# SEL RBn
:SEL rbs_op_n is opcode=0x61; op_l=0xF & rbs_op_n & mov1_sreg=1 [
    rsb = rbs_op_n;
    globalset(inst_next, rsb);
] {}

# NOP
:NOP is opcode=0x00 {}

# HALT
:HALT is opcode=0x61; opcode=0xED { halt(); }

# STOP
:STOP is opcode=0x61; opcode=0xFD { stop(); }

# PREFIX
:^instruction is skip_check=1 & instruction {
    if (SKIP_REG) goto inst_next;
    build instruction;
}

# PREFIX
:^instruction is skip_check=0 & instruction {
    build instruction;
}

# PREFIX
:^instruction is skip_check=0 & es_prefix=0 & opcode=0x11; epsilon ... & instruction [ es_prefix=1; ] {
    build instruction;
}

# PREFIX
:^instruction is skip_check=1 & es_prefix=0 & opcode=0x11; epsilon ... & instruction [ es_prefix=1; ] {
    if (SKIP_REG) goto inst_next;
    build instruction;
}
```

Now with a fresh program I go to a SKC followed by a BR and this is the pcode for it:
```
CMP        A,#0x5
                     $U7100:1 = INT_AND PSW, 0xbf:1
                     $U7080:1 = INT_EQUAL A0, 5:1
                     $U7180:1 = INT_LEFT $U7080:1, 6:4
                     PSW = INT_OR $U7100:1, $U7180:1
                     $U7280:1 = INT_AND PSW, 0xfe:1
                     $U7200:1 = INT_LESS A0, 5:1
                     PSW = INT_OR $U7280:1, $U7200:1
                     $U5180:1 = INT_AND PSW, 0xef:1
                     $U5000:1 = INT_AND A0, 15:1
                     $U5080:1 = INT_AND 5:1, 15:1
                     $U5100:1 = INT_LESS $U5000:1, $U5080:1
                     $U5200:1 = INT_LEFT $U5100:1, 4:4
                     PSW = INT_OR $U5180:1, $U5200:1
SKC
                     SKIP_REG = INT_AND PSW, 1:1
BR         $!LAB_ram_012d54
                     BRANCH *[ram]0x12d54:3
```

There is no conditional branch being emitted. Looking at the instruction info for the `BR` I can see the context_reg is clearly set:
![image](https://user-images.githubusercontent.com/46897303/171510998-eb3f5f41-d6d5-4753-8710-eea2aa4858a6.png)


I'm not sure if this is caused by having multiple `:^instruction ...` or if it is due to something else.

**Reply by GhidorahRex** [reply]
If you look at the instruction info you shared, skip_check is being set, but it's not matching on any of the ^instruction lines. I would make sure those are all moved to the top (although I don't think order matters here).

**Reply by astrelsky** [reply]
> 

I could not get it to work. Order definitely matters though as moving any `:^instruction ...` to the top causes every instruction to fail to be parsed. The reason it fails is it looks at possible matches, checks the first one which is `:^instruction ...` sees it matches, takes it, it builds `instruction`, repeat until maximum depth reached. No matter what order I put it in or how I tweaked it, it wouldn't cooperate.

Edit: I could probably use another bit in the context register to prevent that.

**Reply by astrelsky** [reply]
So changing it to this did work. However, it yielded the same "Multiple flows produced inconsistent instruction prototype ... possibly due to inconsistent context"

These are defined first;

```
# PREFIX
:^instruction is ctx1=0 & skip_check=1 & instruction [ctx1=1;] {
    if (skipflag) goto inst_next;
    build instruction;
}

# PREFIX
:^instruction is ctx1=0 & skip_check=0 & instruction [ctx1=1;] {
    build instruction;
}
```

These are defined later:

```
# SKC
:SKC is opcode=0x61; opcode=0xC8 [skip_check = 1; globalset(inst_next, skip_check);] {
    skipflag = $(CY);
}

# SKNC
:SKNC is opcode=0x61; opcode=0xD8 [skip_check = 1; globalset(inst_next, skip_check);] {
    skipflag = !$(CY);
}

# SKZ
:SKZ is opcode=0x61; opcode=0xE8 [skip_check = 1; globalset(inst_next, skip_check);] {
    skipflag = $(Z);
}

# SKNZ
:SKNZ is opcode=0x61; opcode=0xF8 [skip_check = 1; globalset(inst_next, skip_check);] {
    skipflag = !$(Z);
}

# SKH
:SKH is opcode=0x61; opcode=0xE3 [skip_check = 1; globalset(inst_next, skip_check);] {
    skipflag = ($(Z) | $(CY));
}

# SKNH
:SKNH is opcode=0x61; opcode=0xF3 [skip_check = 1; globalset(inst_next, skip_check);] {
    skipflag = (!($(Z) | $(CY)));
}

# PREFIX
:^instruction is skip_check=0 & es_prefix=0 & opcode=0x11; epsilon ... & instruction [ es_prefix=1; ] {
    build instruction;
}

# PREFIX
:^instruction is skip_check=1 & es_prefix=0 & opcode=0x11; epsilon ... & instruction [ es_prefix=1; ] {
    if (skipflag) goto inst_next;
    build instruction;
}
```

**Reply by GhidorahRex** [reply]
Do you have all the context bits set as noflow?

**Reply by astrelsky** [reply]
> Do you have all the context bits set as noflow?

The ones in use are set as noflow.

**Reply by ghidra1** [reply]
Context definition (flow vs no-flow) is not the issue - it is the differing context for multiple flows leading to the same instruction.  (see comment below)
When differing context occurs this means you are sensitive to the order of disassembly and flow processing which must be avoided since the resulting static pcode would differ.  That is why you must rely on fake registers and dynamic behavior instead of static context to control skipping behavior - or rely on delay slots if they make sense, although they too have their limitations.

**Comment by ghidra1** [comment]
Have you looked at the PA-RISC approach or considered the use of delays slots (you are allowed to branch into a delay slot)?  As you discovered context is dicey for this type of skipping logic and should probably be avoided due to the dynamic nature (an instruction may be skipped or not-skipped depending on flow).

**Reply by astrelsky** [reply]
I don't think you can branch into a delay slot but I might be mistaken. The original implementation used a delay slot and I did away with it for some reason or another. I will definitely have a look at the pa-risc implementation.

**Reply by astrelsky** [reply]
I remember why I did away with the delayslot now. The problem occurs when you have a branch instruction in the delayslot. The flow from that branch isn't followed and disassembly stops. The decompiler just ends up creating `code **` and stuff like it does when it fails to resolve jump tables except there is no warning generated. Otherwise this would be the way to go.

**Reply by ghidra1** [reply]
@astrelsky I don't believe that following the branch flow with a delay slot should be an issue.  The only limitation should be no nesting of delays slotted instructions.  Please open a ticket with details and example of delayslot flow issue.  I did just resolve an issue with context within a delay slot if you are relying on that.

**Reply by astrelsky** [reply]
> @astrelsky I don't believe that following the branch flow with a delay slot should be an issue.  The only limitation should be no nesting of delays slotted instructions.  Please open a ticket with details and example of delayslot flow issue.  I did just resolve an issue with context within a delay slot if you are relying on that. 

Ok, I will open one tonight. I wouldn't expect it to handle this case though since delay slots are associated with branching/jumping/calling. Afaik, at least on mips, you can't have a branch in a delay slot.

**Reply by ghidra1** [reply]
True - but you can have other non-delayslot instructions within a delayslot and branch directly to them.  A branch in a delayslot for mips would cause an attempt to produce a nested delayslot which would fail to disassemble (not permitted).  Sorry, I may have confused this point above.  I was helping another user where they were putting a branch in a delayslot but the branch itself did not have a delaysot (see #4259).

**Reply by astrelsky** [reply]
> True - but you can have other non-delayslot instructions within a delayslot and branch directly to them. A branch in a delayslot for mips would cause an attempt to produce a nested delayslot which would fail to disassemble (not permitted). Sorry, I may have confused this point above. I was helping another user where they were putting a branch in a delayslot but the branch itself did not have a delaysot (see #4259).

Is it possible to override the default disassembler for a processor? If it is then I think the best route is to leave things as it is and have the disassembler set the context variable for the next instruction length as the skip instructions are disassembled instead of during analysis or import.

**Reply by astrelsky** [reply]
> > True - but you can have other non-delayslot instructions within a delayslot and branch directly to them. A branch in a delayslot for mips would cause an attempt to produce a nested delayslot which would fail to disassemble (not permitted). Sorry, I may have confused this point above. I was helping another user where they were putting a branch in a delayslot but the branch itself did not have a delaysot (see #4259).
> 
> Is it possible to override the default disassembler for a processor? If it is then I think the best route is to leave things as it is and have the disassembler set the context variable for the next instruction length as the skip instructions are disassembled instead of during analysis or import.

This is what I have done and it seems to work really well. I'm satisfied with this.

**Reply by ghidra1** [reply]
We do not recommend a custom disassembler.  In addition, this will not solve the case where you branch directly to the optionally skipped-instruction.  This approach also has the same sensitivity to the order that flows are disassembled.

**Reply by ghidra1** [reply]
You are best using fake registers which can be used to control skipping behavior.  If this is done with proper default register values decompiler should be able to simplify.  I believe this was the goal for PA-RISC.

**Reply by astrelsky** [reply]
> We do not recommend a custom disassembler.  In addition, this will not solve the case where you branch directly to the optionally skipped-instruction.  This approach also has the same sensitivity to the order that flows are disassembled. 

I think I just explained poorly. I leveraged the ability to use a custom one and am really just cheating to set context correctly before disassembly. When an instruction is about to be disassembled it sets a context register on the skip instruction. This context register value restricts which of a set of tables are selected that pre-calculates the branch address. The instruction being skipped is just a normal instruction and can be branched to without a problem. Any flow to a skip instruction will attempt to disassemble it will cause the context to be properly set and it will be correctly disassembled. There is a default do nothing method I'm overriding in the disassembler class that appears to be for exactly this use case.



If you do insist I will have another look at the pa risc implementation.

**Reply by astrelsky** [reply]
@ghidra1 I finally got back to this and set it up so that every instruction checks for a skip flag and becomes a NOP if set. The problem I have now is that all `RET` (return) instruction are conditional. This causes the decompiler to go mad and assume the current function is never ending and decompile until it errors out.

**Reply by ghidra1** [reply]
If the conditional skip evaluates to a constant the decompiler should treat as unconditional branch within its simplification.

**Reply by astrelsky** [reply]
> If the conditional skip evaluates to a constant the decompiler should treat as unconditional branch within its simplification.

It might be something else then. The disassembler sees a conditional return and just keeps disassembling into the next function. I was able to get decompiler output after clearing everything in the program but had to scroll past a ton of "Warning: removing unreachable block xyz" to get to the start of the function.

**Reply by ghidra1** [reply]
Yea - not sure how to avoid the disassembler fall-through issue.  Clearly the context approach has issues with multi-flows as you first pointed out.   I have been down this road before and I don't recall a good solution.  What we really need is something similar to `inst_next` which corresponds to the instruction after the next instruction (e.g., `inst_next2`).  I will try to spur discussion on the subject of a possible `inst_next2`.

**Reply by ghidra1** [reply]
I have taken a shot at adding support for a Sleigh `inst_next2` value which can be used to conditionally branch around the next variable length instruction.  At this point I have only modified the Java version of the Sleigh compiler.  I am attaching the code modifications (sorry - they include some long overdue housecleaning of some dead code) as a patch file which can be applied to the latest master branch [03bd4b4](https://github.com/NationalSecurityAgency/ghidra/commit/03bd4b45d00c2bdd46632b01408de2544d87287c).  After applying the patch you can run the command `gradle generateGrammarSource` to generate the Java version of the Sleigh compiler code.
[inst_next2.patch.txt](https://github.com/NationalSecurityAgency/ghidra/files/9377990/inst_next2.patch.txt)

Could you please try this patch out with your RL78.  Be sure to do it in a fashion where you can revert the patch changes to allow for git pulls.  The use of `inst_next2` should eliminate the need to use any context or fake register to facilitate the skip behavior.

**Reply by astrelsky** [reply]
> I have taken a shot at adding support for a Sleigh `inst_next2` value which can be used to conditionally branch around the next variable length instruction. At this point I have only modified the Java version of the Sleigh compiler. I am attaching the code modifications (sorry - they include some long overdue housecleaning of some dead code) as a patch file which can be applied to the latest master branch [03bd4b4](https://github.com/NationalSecurityAgency/ghidra/commit/03bd4b45d00c2bdd46632b01408de2544d87287c). After applying the patch you can run the command `gradle generateGrammarSource` to generate the Java version of the Sleigh compiler code. [inst_next2.patch.txt](https://github.com/NationalSecurityAgency/ghidra/files/9377990/inst_next2.patch.txt)
> 
> Could you please try this patch out with your RL78. Be sure to do it in a fashion where you can revert the patch changes to allow for git pulls. The use of `inst_next2` should eliminate the need to use any context or fake register to facilitate the skip behavior.

Thank you! I will give this is shot tomorrow night/saturday morning and we'll see what I can break.

**Reply by ghidra1** [reply]
I am finding some things I missed so will follow-up with another patch when done.  In addition the decompiler has not yet been changed (C++), although it will probably work OK since it processes pcode and not pcode templates.  I am working on the C++ code now.

**Reply by astrelsky** [reply]
> I am finding some things I missed so will follow-up with another patch when done.  In addition the decompiler has not yet been changed (C++), although it will probably work OK since it processes pcode and not pcode templates.  I am working on the C++ code now.

I suspect the decompiler will work without a problem as long as the c++ sleigh portion is done. The condition branch target would already be computed in the pcode cbranch op once it reaches the rule portion of the decompiler. (This is super rough because I'm going off memory)

**Reply by ghidra1** [reply]
I have attached an update patch file which includes changes to C++ code.  I have not tested the C++ change but it was pretty straight forward.  You may notice places in the code where `inst_next2` indicates "not supported" (e.g., injection, stand-alone decompiler, etc.).  These will not impact your use with Ghidra and may be improved upon in the future.

[inst_next2_patch2.txt](https://github.com/NationalSecurityAgency/ghidra/files/9382100/inst_next2_patch2.txt)

Let me know how it goes...

You will need to build both java sleigh parser and decompiler C++ code:
`gradle generateGrammarSource Decompiler:buildNatives`

**Reply by astrelsky** [reply]
It seems to work good. The only problem I'm seeing is that the disassembler isn't considering the flow when skipping an unconditional branch.

```
SKZ
BR LAB_xyz
// instructions below are not disassembled but the ones at LAB_xyz are
```


I do see the following error in the decompiler on occasion. I'm not sure if it is due to these changes or if it is due to the recent protocol changes.

`Pcode: Decoding error: Expecting <doc> but did not scan an element`

**Reply by ghidra1** [reply]
Here is an updated patch where hopefully the disassembly issue is addressed
[inst_next2_patch4.txt](https://github.com/NationalSecurityAgency/ghidra/files/9397032/inst_next2_patch4.txt)

**Reply by astrelsky** [reply]
It looks like that did the trick.

**Comment by ghidra1** [comment, marked as answer]
Sleigh support for `inst_next2` has been merged into master for 10.2 release.
(See ca80be44ff8ba97da9bb0ead201ef36d5a613795)
